### ISSUE TITLE
how do i get the pruned model?

### ISSUE THREAD

**Issue by SHUSHENGQIGUI** [Skhaki18/optin-transformer-pruning#3]
hello, I am reproducing your excellent work.  however,  i did find the code for saving pruned model weights and architecture. could you tell me how i get the pruned model or provide the relative code?  Thank you!

**Comment by Skhaki18** [comment]
Thank you for your interest in our work. You can find our pre-computed parameter rankings for some of the models on the linked Huggingface page. Once downloaded, and placed in the correct location, per our file tree, you can run our main script and it will evaluate the pruned model. This is completed by loading the pre-computed parameter saved rankings and generating the neuron or attention head masks to be applied to the particular layers on the forward pass using forward hooks. We didn’t need to materialize the static pruned model for our purposes, as FLOPs/performance and our other metrics could be computed directly using the masks. 

If you would like to obtain a statically pruned model, perhaps for additional profiling, you can modify the evaluation scripts for language or vision. In particular, one of the inputs to our `ModelHooking()` object is the `pruningParams`, which at evaluation time represent binarized masks for each layer’s neurons and heads. You may use these to index-select the weights of the model and obtain a statically reduced model architecture.

Hope this helps!

**Comment by SHUSHENGQIGUI** [comment]
> Thank you for your interest in our work. You can find our pre-computed parameter rankings for some of the models on the linked Huggingface page. Once downloaded, and placed in the correct location, per our file tree, you can run our main script and it will evaluate the pruned model. This is completed by loading the pre-computed parameter saved rankings and generating the neuron or attention head masks to be applied to the particular layers on the forward pass using forward hooks. We didn’t need to materialize the static pruned model for our purposes, as FLOPs/performance and our other metrics could be computed directly using the masks.
> 
> If you would like to obtain a statically pruned model, perhaps for additional profiling, you can modify the evaluation scripts for language or vision. In particular, one of the inputs to our `ModelHooking()` object is the `pruningParams`, which at evaluation time represent binarized masks for each layer’s neurons and heads. You may use these to index-select the weights of the model and obtain a statically reduced model architecture.
> 
> Hope this helps!

YES.   i‘d like to obtain a statically pruned model. Thank you, i' ll give it a try

**Comment by SHUSHENGQIGUI** [comment]
> Thank you for your interest in our work. You can find our pre-computed parameter rankings for some of the models on the linked Huggingface page. Once downloaded, and placed in the correct location, per our file tree, you can run our main script and it will evaluate the pruned model. This is completed by loading the pre-computed parameter saved rankings and generating the neuron or attention head masks to be applied to the particular layers on the forward pass using forward hooks. We didn’t need to materialize the static pruned model for our purposes, as FLOPs/performance and our other metrics could be computed directly using the masks.
> 
> If you would like to obtain a statically pruned model, perhaps for additional profiling, you can modify the evaluation scripts for language or vision. In particular, one of the inputs to our `ModelHooking()` object is the `pruningParams`, which at evaluation time represent binarized masks for each layer’s neurons and heads. You may use these to index-select the weights of the model and obtain a statically reduced model architecture.
> 
> Hope this helps!


Hello, I have completed the code for obtaining the reduced model from the language model. However, I encountered two issues while pruning the vision model: 
1. How do I perform patch pruning based on the already obtained patch mask? ( i guess it is token pruning which is different from head pruning or intermediate pruning)
2. When pruning the intermediate layers, I encountered a situation where the neuron mask is all zeros. In this case, the entire layer would be pruned, which would destroy the model structure. How should I handle this?
Could you provide some code examples or more detailed hints?  Thank you!!

**Comment by mariam-jamal** [comment]
Hi, 
@Skhaki18 
I am also struggling to get the static pruned model, could you please guide on how to obtain it? 
What does the patch_mask signify?

**Comment by SHUSHENGQIGUI** [comment]
> Hi, @Skhaki18 I am also struggling to get the static pruned model, could you please guide on how to obtain it? What does the patch_mask signify?

hi i've successfully devoloped the code and completed  patch pruning of vit model.

**Comment by mammadmaheri7** [comment]
> hi i've successfully devoloped the code and completed patch pruning of vit model.

Nice work! Could you share the code for the modifications you made? It would be really helpful.

**Comment by Skhaki18** [comment]
> > Hi, @Skhaki18 I am also struggling to get the static pruned model, could you please guide on how to obtain it? What does the patch_mask signify?
> 
> hi i've successfully devoloped the code and completed patch pruning of vit model.

Hello, This is great news. If you wouldn't mind, could you open a PR for this on the main branch? I can review and merge it into the main code base so others can leverage the same code if needed :)

**Comment by Skhaki18** [comment]
For reference, in our default configuration (beta), the model uses forward hooks to mask out the activations simulating the effect of pruning, hence by indexing into the weights, the mask can be applied and the weight can be statically pruned. In our tau configuration (patch), we incorporated token merging on the fly, to further reduce the FLOPs, however, this does not modify the static size of the model architecture (i.e this will reduce the number of tokens in the forward pass, not the weights). I just patched the code with the token merging implementation we had used. We use `modPatch` as the reduction schedule to apply on ToME.
https://github.com/Skhaki18/optin-transformer-pruning/blob/6c8d7caef2193a48c766f31e56008992dfc0c3bd/evals/vision/eval_imagenet.py#L25

**Comment by SHUSHENGQIGUI** [comment]
> Hi, @Skhaki18 I am also struggling to get the static pruned model, could you please guide on how to obtain it? What does the patch_mask signify?

hi, for head pruning and intermediate  pruning， You can refer to the code of cofi[https://github.com/princeton-nlp/CoFiPruning/blob/main/utils/cofi_utils.py](url), and for patch pruning, it's token pruning actually, so you need to modify the vit forward function to prune token of the input tensor in each encoder layer.

**Comment by SHUSHENGQIGUI** [comment]
> > > Hi, @Skhaki18 I am also struggling to get the static pruned model, could you please guide on how to obtain it? What does the patch_mask signify?
> > 
> > 
> > hi i've successfully devoloped the code and completed patch pruning of vit model.
> 
> Hello, This is great news. If you wouldn't mind, could you open a PR for this on the main branch? I can review and merge it into the main code base so others can leverage the same code if needed :)

of course!

**Comment by Skhaki18** [comment]
> > Hi, @Skhaki18 I am also struggling to get the static pruned model, could you please guide on how to obtain it? What does the patch_mask signify?
> 
> hi, for head pruning and intermediate pruning， You can refer to the code of cofi[https://github.com/princeton-nlp/CoFiPruning/blob/main/utils/cofi_utils.py](url), and for patch pruning, it's token pruning actually, so you need to modify the vit forward function to prune token of the input tensor in each encoder layer.

Yes, this should help. For the patch pruning, it is important to specify that our method determines a reduction scheme for patches as referenced above, hence the number of patches to remove at each layer. However, the methods of removal whether pruning or merging are open to the end user (in our implementation we use ToME). Please feel free to ask for more clarifications if needed, but perhaps @SHUSHENGQIGUI 's pr will help :)
